discovery-provider/src/solana/solana_program_indexer.py

```diff
@@ -0,0 +1,263 @@
+        for tx_sig in tx_sig_batch_records:
+            parsed_transactions.append(tx_sig_futures_map[tx_sig])
+
+        # Fetch metadata in parallel
+        metadata_dictionary = self.fetch_ipfs_metadata(parsed_transactions)
+
+        self.validate_and_save_parsed_tx_records(
+            parsed_transactions, metadata_dictionary
+        )
+
+    @abstractmethod
+    def validate_and_save_parsed_tx_records(
+        self, parsed_transactions, metadata_dictionary
+    ):
+        """
+        Based parsed transaction information, generate and save appropriate database changes. This will vary based on the program being indexed
+        @param parsed_transactions: Array of transaction signatures in order
+        @param metadata_dictionary: Dictionary of remote metadata
+        """
+        raise Exception(BASE_ERROR)
+
+    def get_transactions_to_process(self):
+        """
+        Calculate the delta between database and chain tail and return an array of arrays containing transaction batches
+        """
+        latest_processed_slot = self.get_latest_slot()
+        # The 'before' value from where we start querying transactions
+        # TODO: Add cache
+        last_tx_signature = None
+
+        # Loop exit condition
+        intersection_found = False
+
+        # List of signatures that will be populated as we traverse recent operations
+        transaction_signatures = []
+
+        # Current batch of transactions
+        transaction_signature_batch = []
+
+        # Current batch
+        page_count = 0
+        while not intersection_found:
+            self.msg(
+                f"Requesting transactions before {last_tx_signature} - {self._program_id}"
+            )
+            transactions_history = (
+                self._solana_client_manager.get_signatures_for_address(
+                    self._program_id,
+                    before=last_tx_signature,
+                    limit=TX_SIGNATURES_BATCH_SIZE,
+                )
+            )
+            transactions_array = transactions_history["result"]
+            self.msg(f"{transactions_array}")
+            if not transactions_array:
+                # This is considered an 'intersection' since there are no further transactions to process but
+                # really represents the end of known history for this ProgramId
+                intersection_found = True
+                self.msg(f"No transactions found before {last_tx_signature}")
+            else:
+                with self._db.scoped_session() as read_session:
+                    for tx in transactions_array:
+                        tx_sig = tx["signature"]
+                        slot = tx["slot"]
+                        if tx["slot"] > latest_processed_slot:
+                            transaction_signature_batch.append(tx_sig)
+                        elif tx["slot"] <= latest_processed_slot:
+                            # Check the tx signature for any txs in the latest batch,
```

Comment:
w/ the block of code underneath this, would strongly prefer to wrap this logic in functions instead of the comments - would help a lot w readability and typing/scoping. we can break out find_intersection into its own fn, for instance.

Reply:
agreed, I will folllow up here